src/transaction_queue.rs

```diff
@@ -1,32 +1,56 @@
-use std::cmp;
-use std::collections::{HashSet, VecDeque};
+use std::collections::HashSet;
+use std::{cmp, fmt};
 
-use rand;
+use rand::{self, Rng};
 
 use Contribution;
 
-/// A wrapper providing a few convenience methods for a queue of pending transactions.
-#[derive(Debug)]
-pub struct TransactionQueue<T>(pub VecDeque<T>);
-
-impl<T: Clone> TransactionQueue<T> {
+/// An interface to the transaction queue. A transaction queue is a structural part of
+/// `QueueingHoneyBadger` that manages enqueueing of transactions for a future batch and dequeueing
+/// of transactions to become part of a current batch.
+pub trait TransactionQueue<T>: fmt::Debug + Default + Extend<T> + Sync + Send {
+    /// Checks whether the queue is empty.
+    fn is_empty(&self) -> bool;
+    /// Appends an element at the end of the queue.
+    fn push(&mut self, t: T);
```

Comment:
https://play.rust-lang.org/?version=stable&mode=release&edition=2015  --  Comparing the ASM output, it's almost exactly the same, except that operands are flipped on a single `cmpq` instruction =).

I still think we should use extend here and/or provide a default impl.